libres/lib/analysis/update.cpp

```diff
@@ -763,7 +783,28 @@ bool smoother_update(std::vector<int> step_list,
     int_vector_free(ens_active_list);
     meas_data_free(meas_data);
     bool_vector_free(ens_mask);
-
+    fclose(log_stream);
     return true;
 }
 } // namespace analysis
+
+static bool smoother_update(std::vector<int> step_list, py::object updatestep,
```

Comment:
I neglected to mention this, but ideally all vectors should be made `const&` in function parameters. It sucks that it's more verbose, but it's good form.

```suggestion
static bool smoother_update(const std::vector<int>& step_list, py::object updatestep,
```

Reason:
```c++
// C++ uses value semantics by default
std::vector<int> a = big_vector();
std::vector<int> b = a;  /* Make a copy of 'a' and put it into 'b'. Potentially very expensive */

// But we need to prefer reference semantics
std::vector<int> a = big_vector();
std::vector<int>& b = a;  /* 'b' is semantically 'a'. No copy is made. */
```

```py
# Python is pass-by-reference and doesn't have value semantics.
# This is more like pointers in C or `object` type in Java.
a = big_vector()
b = a  # a refers to the same object as b
assert b is a  # => true
```

C++ references are a bit more funky since you cannot distinguish between a reference and the object it refers to. This makes it possible for the compiler to implement references as pointers or just rename `b` to `a` since it may have additional knowledge. Eg. taking the address of a reference `&b` will give you the address of the original object. So `&b == &a`. Even though `b` could've been implemented as a pointer in the machine code.

Anyway. The rule of thumb is that you do `const&` for everything, except anything that is made cheap to copy: all primitives (`int`, `double`, `T *`), `std::string_view` (`const std::string&` is still the correct choice), function pointers, and `py::handle`/`py::object`.

Reply:
I agree